include/bluetooth/audio/vcs.h

```diff
@@ -0,0 +1,504 @@
+#endif
+
+#if defined(CONFIG_BT_VCS)
+#define BT_VCS_VOCS_CNT CONFIG_BT_VCS_VOCS_INSTANCE_COUNT
+#define BT_VCS_AICS_CNT CONFIG_BT_VCS_AICS_INSTANCE_COUNT
+#else
+#define BT_VCS_VOCS_CNT 0
+#define BT_VCS_AICS_CNT 0
+#endif /* CONFIG_BT_VCS */
+
+/** Volume Control Service Error codes */
+#define BT_VCS_ERR_INVALID_COUNTER             0x80
+#define BT_VCS_ERR_OP_NOT_SUPPORTED            0x81
+
+/** Register structure for Volume Control Service */
+struct bt_vcs_register_param {
+	/** Register parameters for Volume Offset Control Services */
+	struct bt_vocs_register_param vocs_param[BT_VCS_VOCS_CNT];
+
+	/** Register parameters  for Audio Input Control Services */
+	struct bt_aics_register_param aics_param[BT_VCS_AICS_CNT];
+
+	/** Volume Control Service callback structure. */
+	struct bt_vcs_cb *cb;
+};
+
+/**
+ * @brief Volume Control Service service instance
+ *
+ * Used for to represent a Volume Control Service instance, for either a client
+ * or a server. The instance pointers either represent local server instances,
+ * or remote service instances.
+ */
+struct bt_vcs {
+	/** Number of Volume Offset Control Service instances */
+	uint8_t vocs_cnt;
+	/** Array of pointers to Volume Offset Control Service instances */
+	struct bt_vocs **vocs;
+
+	/** Number of Audio Input Control Service instances */
+	uint8_t aics_cnt;
+	/** Array of pointers to Audio Input Control Service instances */
+	struct bt_aics **aics;
+};
+
+/**
+ * @brief Register the Volume Control Service.
+ *
+ * This will register and enable the service and make it discoverable by
+ * clients.
+ *
+ * @param param     Volume Control Service register parameters.
+ *
+ * @return 0 if success, errno on failure.
+ */
+int bt_vcs_register(struct bt_vcs_register_param *param);
+
+/**
+ * @brief Get Volume Control Service service pointer.
+ *
+ * Returns a pointer to a struct that contains information about the
+ * Volume Control Service instance, such as pointers to the
+ * Volume Offset Control Service (Volume Offset Control Service) or
+ * Audio Input Control Service (AICS) instances.
+ *
+ * @param conn          Connection to peer device, or NULL to get server value.
+ * @param[out] service  Pointer to store the result in.
+ *
+ * @return 0 if success, errno on failure.
+ */
+int bt_vcs_get(struct bt_conn *conn, struct bt_vcs *service);
+
+/**
+ * @brief Callback function for bt_vcs_discover
+ *
+ * @param conn         The connection that was used to discover
+ *                     Volume Control Service.
+ * @param err          Error value. 0 on success, GATT error on positive value
+ *                     or errno on negative value.
+ * @param vocs_count   Number of Volume Offset Control Service instances
+ *                     on peer device.
+ * @param aics_count   Number of Audio Input Control Service instances on
+ *                     peer device.
+ */
+typedef void (*bt_vcs_discover_cb)(struct bt_conn *conn, int err,
+				   uint8_t vocs_count, uint8_t aics_count);
```

Comment:
@asbjornsabo Any suggestions on how to? 

One suggestion I have, is to refactor the callback struct to be similar to `struct bt_conn_cb` where the callbacks are defined inside the struct, and "special" cases are guarded by ifdefs, so e.g. the vcs callback structure would be something like

```
struct bt_vcs_cb {
	void (*state_changed_cb)(struct bt_conn *conn, int err, uint8_t volume, uint8_t mute);
#if defined(CONFIG_BT_VCS_CLIENT)
	void (*discover_cb)(struct bt_conn *conn, int err, uint8_t vocs_count, uint8_t aics_count);
#endif /* CONFIG_BT_VCS_CLIENT */
};
```

Reply:
I had a more naive thought - just mention it in the Doxygen documentation.
Since the rule for most functions is that they apply on both sides, it should be sufficient to note the exceptions, e.g. like 
" * (Client side only)"

Reply:
I've added that for all server or client specific functions and callbacks.